packages/electron/src/electron-main/electron-application.ts

```diff
@@ -0,0 +1,388 @@
+        const electronWindow = new BrowserWindow(options);
+        // Prevent openning arbitrary links in BrowserWindow instances.
+        // A link is to be displayed in the user's default browser.
+        electronWindow.webContents.on('new-window', (event, url) => {
+            event.preventDefault();
+            shell.openExternal(url);
+        });
+
+        electronWindow.on('ready-to-show', () => electronWindow.show());
+        this.attachSaveWindowState(electronWindow);
+        this.attachWillPreventUnload(electronWindow);
+
+        return electronWindow;
+    }
+
+    async openDefaultWindow(): Promise<BrowserWindow> {
+        const uri = await this.createWindowUri();
+        const window = await this.createWindow(this.getBrowserWindowOptions());
+        window.loadURL(uri.toString(true));
+        return window;
+    }
+
+    async openWindowWithWorkspace(workspace: string): Promise<BrowserWindow> {
+        const uri = (await this.createWindowUri()).withFragment(workspace);
+        const window = await this.createWindow(this.getBrowserWindowOptions());
+        window.loadURL(uri.toString(true));
+        return window;
+    }
+
+    /**
+     * "Gently" close all windows, application will not stop if a `beforeunload`
+     * handler returns `false`.
+     */
+    requestStop(): void {
+        app.quit();
+    }
+
+    protected async handleMainCommand(params: ExecutionParams, options: MainCommandOptions): Promise<void> {
+        if (typeof options.file === 'undefined') {
+            await this.openDefaultWindow();
+        } else {
+            await this.openWindowWithWorkspace(realpathSync(path.resolve(params.cwd, options.file)));
+        }
+    }
+
+    protected async createWindowUri(): Promise<URI> {
+        return new URI()
+            .withScheme('file')
+            .withPath(this.settings.THEIA_FRONTEND_HTML_PATH)
+            .withQuery(`port=${this.backendPort}`);
+    }
+
+    protected getBrowserWindowOptions(): BrowserWindowConstructorOptions {
+        let windowState: BrowserWindowConstructorOptions | undefined = this.electronStore.get('windowstate', undefined);
+        if (typeof windowState === 'undefined') {
+            windowState = this.getDefaultWindowState();
+        }
+        return {
+            ...windowState,
+            show: false,
+            title: this.settings.THEIA_APPLICATION_NAME,
+            minWidth: 200,
+            minHeight: 120,
+        };
+    }
+
+    protected getDefaultWindowState(): BrowserWindowConstructorOptions {
+        // The `screen` API must be required when the application is ready.
+        // See: https://electronjs.org/docs/api/screen#screen
+        const { screen } = require('electron');
+
+        // We must center by hand because \`browserWindow.center()\` fails on multi-screen setups
+        // See: https://github.com/electron/electron/issues/3490
+        const { bounds } = screen.getDisplayNearestPoint(screen.getCursorScreenPoint());
```

Comment:
Are you saying the [logic from the `HEAD` of the `master`](https://github.com/eclipse-theia/theia/blob/310ff528da74e6c8ccb7abe403040e82cfc73fb3/dev-packages/application-manager/src/generator/frontend-generator.ts#L178-L182) is incorrect? If so, can we handle it in a follow-up? If no, let's not change it.

Reply:
Regarding the code you highlighted, I just moved the code out of the generators and into the inversified classes almost verbatim.